wallet/src/de/schildbach/wallet/ui/main/MainViewModel.kt

```diff
@@ -524,201 +500,223 @@ class MainViewModel @Inject constructor(
+                    val rowView = TransactionRowView.fromTransactionWrapper(
                         it,
                         walletData.transactionBag,
                         Constants.CONTEXT,
-                        contact,
-                        metadata[it.transactions.first().txId]
+                        null,
+                        metadata[tx.txId]
                     )
+                    txByHash[rowView.id] = rowView
+                    rowView
+                }
+                .groupBy {
+                    Instant.ofEpochMilli(it.time).atZone(ZoneId.systemDefault()).toLocalDate()
+                }
+                .mapValues { (_, transactions) ->
+                    transactions.sortedByDescending { it.time }
                 }
 
-            transactionViews = allTransactionViews.toMutableList()
-            log.info("refreshTransactions: {} ms", watch.elapsed(TimeUnit.MILLISECONDS))
-            _transactions.postValue(transactionViews)
-        }
-    }
-
-    /** update a batch of transactions */
-    private suspend fun refreshTransactions(
-        txList: List<Transaction>,
-        filter: TxDirectionFilter,
-        metadata: Map<Sha256Hash, PresentableTxMetadata>
-    ) {
-        var updateCount = 0
-        log.info("refreshTransactions({}, but only {} unique tx)", txList.size, txList.toSet().size)
-        txList.toSet().forEach { tx ->
-            if (filter.matches(tx)) {
-                refreshTransaction(tx, transactionViews, filter, metadata)
-                updateCount++
+            viewModelScope.launch {
+                _transactions.value = allTransactionViews
+                this@MainViewModel.txByHash = txByHash
+                updateContacts(contactsToUpdate)
             }
         }
-        if (updateCount > 0) {
-            _transactions.postValue(transactionViews.toList())
-        }
     }
 
     /**
      * this will either add a single transaction or update the transaction on the current view
      */
-    private suspend fun refreshTransaction(
-        tx: Transaction,
-        updatedList: MutableList<TransactionRowView>,
-        filter: TxDirectionFilter, metadata: Map<Sha256Hash, PresentableTxMetadata>) {
+    private fun refreshTransactionBatch(
+        transactions: List<Transaction>,
+        filter: TxDirectionFilter,
+        contacts: Map<Sha256Hash, DashPayProfile>
+    ) {
         Context.propagate(Constants.CONTEXT)
-        val watch = Stopwatch.createStarted()
-        // is the item currently in our list
-        val txRowViewIndex = transactionViews.indexOfFirst {
-            if (it.txId == tx.txId) {
-                true
-            } else {
-                it.txWrapper?.transactions?.find { txInGroup ->
-                    tx.txId == txInGroup.txId
-                } != null
+        val items = _transactions.value!!.toMutableMap()
+        val txByHash = this.txByHash.toMutableMap()
```

Comment:
Keeping exposed fields unmutable and creating mutable collections when modifying should help to avoid concurrent modification errors